src/charm.py

```diff
@@ -0,0 +1,131 @@
+
+    def _rebuild_waltz_pebble_layer(self, container=None):
+        """(Re)Builds the Waltz Pebble Layer.
+
+        The Pebble layer will be (re)created if the credentials required by
+        Waltz to connect to a postgresql database are present. The layer will
+        only be recreated if it is different from the current specification.
+        """
+        # Check if we have all the necessary information to start Waltz.
+        db_config = self._get_database_config()
+        if not db_config:
+            self.unit.status = model.BlockedStatus("Waiting for database configuration.")
+            return
+
+        # If not provided with a container, get one.
+        container = container or self.unit.get_container("waltz")
+
+        if not container.can_connect():
+            self.unit.status = model.WaitingStatus("Waiting for Pebble to be ready")
+            return
+
+        # Create the current Pebble layer configuration
+        pebble_layer = pebble.Layer(self._generate_workload_pebble_layer(db_config))
+
+        plan = container.get_plan()
+        updated_services = plan.services != pebble_layer.services
+        if updated_services:
+            logger.info("Waltz needs to be updated. Restarting.")
+            container.add_layer("waltz", pebble_layer, combine=True)
+            container.restart("waltz")
+
+        self.unit.status = model.ActiveStatus()
+
+    def _generate_workload_pebble_layer(self, db_config):
+        """Generates the Waltz layer for Pebble.
+
+        The generated layer will also contain the database configuration required by FINOS Waltz
+        into the environment.
+        """
+        return {
+            "summary": "Waltz layer",
+            "description": "Pebble config layer for Waltz.",
+            "services": {
+                "waltz": {
+                    "override": "replace",
+                    "summary": "waltz",
+                    "command": "/bin/sh -c 'docker-entrypoint.sh update run'",
+                    "startup": "enabled",
+                    "user": "waltz",
+                    "environment": {
+                        "DB_HOST": db_config["host"],
+                        "DB_PORT": db_config["port"],
+                        "DB_NAME": db_config["dbname"],
+                        "DB_USER": db_config["username"],
+                        "DB_PASSWORD": db_config["password"],
+                        "DB_SCHEME": "waltz",
+                        "WALTZ_FROM_EMAIL": "help@finos.org",
+                        "CHANGELOG_FILE": "/opt/waltz/liquibase/db.changelog-master.xml",
+                    },
+                }
+            },
+        }
+
+    def _on_config_changed(self, _):
+        """Refreshes the service config.
+
+        A new Waltz pebble layer specification will be set if any of the
+        configuration options related to Waltz are updated.
+        """
+        logger.debug("Updated Waltz Charm config.")
```

Comment:
I'd drop this, if debug logging is switched on for the model, Juju logs events being dispatched anyway :)